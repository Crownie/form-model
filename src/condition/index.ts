export {Condition} from './condition';
export {ConditionBuilder} from './condition-builder';
export {ConditionGroupBuilder} from './condition-group-builder';
