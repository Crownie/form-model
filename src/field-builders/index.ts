export * from './boolean-field-builder';
export * from './date-time-field-builder';
export * from './field-builder';
export * from './list-field-builder';
export * from './money-field-builder';
export * from './object-field-builder';
export * from './select-field-builder';
export * from './slider-field-builder';
export * from './text-field-builder';
