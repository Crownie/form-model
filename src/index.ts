export * from './condition';
export * from './field-builders';
export * from './field';
export * from './form';
export * from './wiz-form';
