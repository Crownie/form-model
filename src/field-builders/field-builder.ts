import {
  Field,
  DataType,
  OptionItem,
  ResponsiveSize,
  SchemaConfig,
} from '../field';
import {Condition} from '../condition';

export class FieldBuilder {
  public type: string;
  public label: string;
  public width: ResponsiveSize = {xs: 12};
  public noWrap: boolean = false;
  public childField?: Field;
  public autocomplete?: boolean;
  public fields?: {[key: string]: Field};
  public options?: OptionItem[];
  public format?: string;
  public meta?: Record<string, any> = {};
  public conditionalDisable?: Condition;
  public conditionalHide?: Condition;
  public schemaConfig: SchemaConfig = {
    type: 'mixed',
  };
  private static defaults = {
    string: '',
    number: '',
    boolean: false,
    array: [],
    object: {},
    date: null,
  };

  constructor(fieldType: string, label: string, dataType: DataType = 'string') {
    this.type = fieldType;
    this.label = label;
    this.schemaConfig.type = dataType;
    this.schemaConfig.defaultVal = FieldBuilder.defaults[dataType];
    if (dataType === 'object') {
      this.fields = {};
    }
  }

  getDefaultValue() {
    return this.schemaConfig.defaultVal;
  }

  setWidth(width: ResponsiveSize, noWrap: boolean = false): this {
    this.noWrap = noWrap;
    this.width = {...this.width, ...width};
    return this;
  }

  arrayType(schemaConfig: SchemaConfig) {
    this.schemaConfig.type = 'array';
    this.schemaConfig.arrayType = schemaConfig;
  }

  objectShape(shape: SchemaConfig['objectShape']) {
    this.schemaConfig.type = 'object';
    this.schemaConfig.objectShape = shape;
  }

  getSchemaConfig(): SchemaConfig {
    return this.schemaConfig;
  }

  required(errorMessage: string = 'Required'): this {
    this.schemaConfig.required = {
      val: errorMessage,
      errorMessage,
    };
    return this;
  }

  defaultVal(val: any): this {
    this.schemaConfig.defaultVal = val;
    return this;
  }

  min(val: any, errorMessage: string = 'Too short'): this {
    this.schemaConfig.min = {
      val,
      errorMessage,
    };
    return this;
  }

  max(val: any, errorMessage: string = 'Too long'): this {
    this.schemaConfig.max = {
      val,
      errorMessage,
    };
    return this;
  }

  email(errorMessage: string = 'Not a valid email'): this {
    this.schemaConfig.email = {
      val: errorMessage,
      errorMessage,
    };
    return this;
  }

  matches(val: RegExp, errorMessage: string): this {
    this.schemaConfig.matches = {val, errorMessage};
    return this;
  }

  equalToField(fieldName: string, errorMessage: string): this {
    this.schemaConfig.equalToField = {
      val: fieldName,
      errorMessage,
    };
    return this;
  }

  nullable(): this {
    this.schemaConfig.nullable = true;
    return this;
  }

  addMeta(key: string, value: any): this {
    this.meta[key] = value;
    return this;
  }

  disableWhen(condition: Condition): this {
    this.conditionalDisable = condition;
    return this;
  }

  hideWhen(condition: Condition): this {
    this.conditionalHide = condition;
    return this;
  }

  build(): Field {
    return new Field(this);
  }

  protected setDataType(dataType: DataType) {
    this.schemaConfig.type = dataType;
  }
}
