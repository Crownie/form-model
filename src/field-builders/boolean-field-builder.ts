import {FieldBuilder} from './field-builder';

export class BooleanFieldBuilder extends FieldBuilder {
  constructor(label: string, type: 'switch' | 'checkbox' = 'checkbox') {
    super(type, label, 'boolean');
  }
}
