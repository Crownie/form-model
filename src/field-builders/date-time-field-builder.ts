import {FieldBuilder} from './field-builder';

export class DateTimeFieldBuilder extends FieldBuilder {
  constructor(label: string) {
    super('date-time', label, 'date');
    this.format = 'MM/DD/YYYY hh:mm A';
  }

  dateOnly(format: string = 'MM/DD/YYYY'): this {
    this.type = 'date';
    this.format = format;
    return this;
  }

  timeOnly(format: string = 'hh:mm A'): this {
    this.type = 'time';
    this.format = format;
    return this;
  }
}
