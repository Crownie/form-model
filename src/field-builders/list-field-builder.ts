import {FieldBuilder} from './field-builder';
import {Field} from '../field';

export class ListFieldBuilder extends FieldBuilder {
  constructor(label: string) {
    super('list', label, 'array');
  }

  addField(fieldOrFieldBuilder: Field | FieldBuilder) {
    const field =
      fieldOrFieldBuilder instanceof FieldBuilder
        ? fieldOrFieldBuilder.build()
        : fieldOrFieldBuilder;
    this.childField = field;
    this.arrayType(field.getSchemaConfig());
    return this;
  }
}
