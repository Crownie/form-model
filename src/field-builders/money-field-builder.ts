import {FieldBuilder} from './field-builder';

export class MoneyFieldBuilder extends FieldBuilder {
  constructor(label: string) {
    super('money', label, 'object');
    this.objectShape({
      currency: {
        type: 'string',
      },
      amount: {
        type: 'number',
      },
    });
    this.defaultVal({currency: '', amount: ''});
  }

  required(errorMessage: string = 'Required'): this {
    this.schemaConfig.objectShape!.currency.required = {
      val: errorMessage,
      errorMessage,
    };
    this.schemaConfig.objectShape!.amount.required = {
      val: errorMessage,
      errorMessage,
    };
    return super.required();
  }

  min(val: any, errorMessage: string = `Minimum is ${val}`): this {
    this.schemaConfig.objectShape!.amount.min = {val, errorMessage};
    return this;
  }

  max(val: any, errorMessage: string = `Maximum is ${val}`): this {
    this.schemaConfig.objectShape!.amount.max = {val, errorMessage};
    return this;
  }
}
