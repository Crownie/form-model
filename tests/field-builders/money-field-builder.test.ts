import {MoneyFieldBuilder} from '../../src/field-builders';

it('MoneyFieldBuilder', () => {
  const fieldBuilder = new MoneyFieldBuilder('price')
    .required()
    .min(1)
    .max(100);

  expect(fieldBuilder.build()).toMatchSnapshot();
});
