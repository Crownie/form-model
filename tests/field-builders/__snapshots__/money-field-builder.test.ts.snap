// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoneyFieldBuilder 1`] = `
Field {
  "autocomplete": undefined,
  "childField": undefined,
  "conditionalDisable": undefined,
  "conditionalHide": undefined,
  "defaults": Object {
    "array": Array [],
    "boolean": false,
    "number": "",
    "object": Object {},
    "string": "",
  },
  "fields": Object {},
  "format": undefined,
  "label": "price",
  "noWrap": false,
  "options": undefined,
  "schemaConfig": Object {
    "defaultVal": Object {
      "amount": "",
      "currency": "",
    },
    "objectShape": Object {
      "amount": Object {
        "max": Object {
          "errorMessage": "Maximum is 100",
          "val": 100,
        },
        "min": Object {
          "errorMessage": "Minimum is 1",
          "val": 1,
        },
        "required": Object {
          "errorMessage": "Required",
          "val": "Required",
        },
        "type": "number",
      },
      "currency": Object {
        "required": Object {
          "errorMessage": "Required",
          "val": "Required",
        },
        "type": "string",
      },
    },
    "required": Object {
      "errorMessage": "Required",
      "val": "Required",
    },
    "type": "object",
  },
  "type": "money",
  "width": Object {
    "xs": 12,
  },
}
`;
